docs/enhacement/20201211-nodepool_uniteddeployment.md

```diff
@@ -0,0 +1,374 @@
+    // +optional
+    Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+    // If specified, the Labels will be added to all nodes.
+    // NOTE: existing labels with samy keys on the nodes will be overwritten.
+    // +optional
+    Labels map[string]string `json:"labels,omitempty"`
+
+    // If specified, the Annotations will be added to all nodes.
+    // NOTE: existing labels with samy keys on the nodes will be overwritten.
+    // +optional
+    Annotations map[string]string `json:"annotations,omitempty"`
+
+    // If specified, the Taints will be added to all nodes.
+    // +optional
+    Taints []v1.Taint `json:"taints,omitempty"`
+}
+
+// NodePoolStatus defines the observed state of NodePool
+type NodePoolStatus struct {
+    // Total number of ready nodes in the pool.
+    // +optional
+    ReadyNode int32 `json:"readyNode"`
+
+    // Total number of not ready nodes in the pool.
+    // +optional
+    NotReadyNode int32 `json:"notReadyNode"`
+
+    // The list of nodes' names in the pool
+    // +optional
+    EdgeNodes map[string]bool `json:"edgeNodes,omitempty"`
+
+    // Labels/annotations/taints that are currently used by nodes of the pool
+    // +optional
+    ManagedItems ManagedItems `json:"managedItems,omitempty"`
+}
+
+// ManagedItems contians the labels/annotations/taints that are currently
+// attached to nodes of the pool
+type ManagedItems struct {
+    // +optional
+    Labels map[string]string `json:"labels,omitempty"`
+
+    // +optional
+    Annotations map[string]string `json:"annotations,omitempty"`
+
+    // +optional
+    Taints []v1.Taint `json:"taints,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster,path=nodepools,shortName=np,categories=all
+// +kubebuilder:printcolumn:name="Type",type="string",JSONPath=".spec.type",description="The type of nodepool"
+// +kubebuilder:printcolumn:name="ReadyNodes",type="integer",JSONPath=".status.readyNode",description="The number of ready nodes in the pool"
+// +kubebuilder:printcolumn:name="NotReadyNodes",type="integer",JSONPath=".status.notReadyNode"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +genclient:nonNamespaced
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient
+// NodePool is the Schema for the nodepools API
+type NodePool struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   NodePoolSpec   `json:"spec,omitempty"`
+    Status NodePoolStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// NodePoolList contains a list of NodePool
+type NodePoolList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []NodePool `json:"items"`
+}
+```
+
+
+### UnitedDeployment API
+
+``` go
+
+// UnitedDeploymentConditionType indicates valid conditions type of a UnitedDeployment.
```

Comment:
should we change "conditions type" to "condition type"?

Reply:
Will change to `condition type`